src/cascadia/TerminalSettingsModel/CascadiaSettingsSerialization.cpp

```diff
@@ -432,92 +440,124 @@ gsl::span<const winrt::com_ptr<Profile>> SettingsLoader::_getNonUserOriginProfil
             {
-                if (profile->HasName())
+                if (const auto scheme = ColorScheme::FromJson(schemeJson))
                 {
-                    profile->Guid(profile->Guid());
+                    settings.globals->AddColorScheme(*scheme);
                 }
-                else if (!updatesKeyAllowed || profile->Updates() == winrt::guid{})
+            }
+            CATCH_LOG()
+        }
+    }
+
+    {
+        const auto size = json.profilesList.size();
+        settings.profiles.reserve(size);
+        settings.profilesByGuid.reserve(size);
+
+        for (const auto& profileJson : json.profilesList)
+        {
+            try
+            {
+                auto profile = _parseProfile(OriginTag::Fragment, source, profileJson);
+                if (profile->HasGuid() || profile->Updates() != winrt::guid{})
                 {
-                    // We introduced a bug (GH#9962, fixed in GH#9964) that would result in one or
-                    // more nameless, guid-less profiles being emitted into the user's settings file.
-                    // Those profiles would show up in the list as "Default" later.
-                    //
-                    // Fragments however can contain an alternative "updates" key, which works similar to the "guid".
-                    // If updatesKeyAllowed is true (see FindFragmentsAndMergeIntoUserSettings) we permit
-                    // such Guid-less, Name-less profiles as long as they have a valid Updates field.
-                    continue;
+                    _appendProfile(std::move(profile), settings);
                 }
             }
-
-            _appendProfile(std::move(profile), settings);
+            CATCH_LOG()
         }
     }
 }
 
+SettingsLoader::JsonSettings SettingsLoader::_parseJson(const std::string_view& content)
+{
+    auto root = content.empty() ? Json::Value{ Json::ValueType::objectValue } : _parseJSON(content);
+    const auto& colorSchemes = _getJSONValue(root, SchemesKey);
+    const auto& profilesObject = _getJSONValue(root, ProfilesKey);
+    const auto& profileDefaults = _getJSONValue(profilesObject, DefaultSettingsKey);
+    const auto& profilesList = profilesObject.isArray() ? profilesObject : _getJSONValue(profilesObject, ProfilesListKey);
+    return JsonSettings{ std::move(root), colorSchemes, profileDefaults, profilesList };
+}
+
+// Just a common helper function between _parse and _parseFragment. Parses a profile.
+winrt::com_ptr<Profile> SettingsLoader::_parseProfile(const OriginTag origin, const winrt::hstring& source, const Json::Value& profileJson)
+{
+    auto profile = Profile::FromJson(profileJson);
+    profile->Origin(origin);
+
+    // The Guid() generation below depends on the value of Source().
+    // --> Provide one if we got one.
+    if (!source.empty())
+    {
+        profile->Source(source);
+    }
+
+    if (!profile->HasGuid() && profile->HasName())
+    {
+        profile->Guid(profile->Guid());

```

Comment:
This looks really weird haha. I'd add a comment saying that we're generating a guid based on the name, and storing it.